Fix bloom size issues with Windows serialization and in general

polm opened the pull request:
Before this commit, serialization used the array module, but the size of units in that varies depending on the host OS. As a result data serialized on Windows would use 32-bit units instead of 64-bit ones.

This commit uses `struct` for serialization instead of `array`, eliminating the ambiguity issue. Additionally, it detects if the serialized data used the Windows size, and loads it in that size if necessary, so old data will still work. (Even if the data was serialized into smaller containers on Windows no data was lost due to a separate issue, which will be fixed in a further commit.)

Note that this issue only affected the serialization code, and in memory containers were consistent on platforms already.

This is a draft until the further bloom size issues are addressed.

polm commented:
I believe Windows is working now, though I'm leaving this a draft while I verify the result of serializing data on Windows. 

Going through backwards compatability, one other issue came up - you could create a BloomFilter with size/hash function count params of 0. If the hash count is 0 add/contains checks do nothing. If size is 0 and hcount is not, then it will crash if you try to add anything (unsurprisingly). I added asserts to check for this.

I had one question come up while doing this - what's the right way to declare a const value in Cython? For some of the arithmetic I need the size of a container type in bits, so I'm assigning that at the top of the file.

polm commented:
I verified the Windows data is correct on real Windows.

The test failures here are unexpected - it looks like there is a segfault only on 3.6 on Linux. I'll look into that.

polm commented:
I am unable to reproduce the segfault locally, even with the exact Python version used in the tests (3.6.15). I am not sure what else could explain it, I'll see if I can figure something out.

svlandeg commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,108 @@ cdef class BloomFilter:
+    bloom.seed = seed
+    buflen = length // KEY_BITS
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
```
Whenever calling `mem.alloc`, ensure (either by checking or asserting) that the value is not `0` - cf https://github.com/explosion/spaCy/pull/4423 - just to avoid any nasty bugs down the line ;-)

polm replied:
Thanks for the tip, I added asserts to check that there are no calls to alloc with 0 length. It doesn't seem to be the issue this time though...

svlandeg replied:
Ah that's unfortunate - so we'll need to keep digging ;-)

polm commented on preshed/bloom.pyx:
```diff
@@ -3,13 +3,15 @@
 except ImportError:
     import copyreg as copy_reg
 
+# TODO better way to declare constant?
+cdef key_t KEY_BITS = 8 * sizeof(key_t)
```
Is there a better way to declare a constant? Should I use an anonymous enum [like the Cython docs say](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#:~:text=but%20you%20can%20use%20an%20anonymous%20enum%20declaration%20for%20this%20purpose)?

danieldk replied:
Yes, that is better than a mutable variable.

polm commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    # note that the modulus check is only required for data that has come from
+    # legacy deserialization - otherwise length is always a multiple of
+    # KEY_BITS.
+    buflen = bloom.length // KEY_BITS
+    if bloom.length % KEY_BITS > 0:
+        buflen += 1
```
I use this pattern for ceiling division in several places, but I guess upside-down floor division would be better? That would look like this:

```
buflen = -(bloom.length // -KEY_BITS)
```

It's more terse, and perhaps faster, but also rather weird looking.

polm commented:
There's a few things I'd like feedback on, but this is functional and has no issues I'm aware of at this point.

danieldk commented on preshed/bloom.pyx:
```diff
@@ -109,8 +198,8 @@ cdef void bloom_add(BloomStruct* bloom, key_t item) nogil:
-        hv = (keys[0] + (hiter * keys[1])) % bloom.length
-        bloom.bitfield[hv // sizeof(key_t)] |= one << (hv % sizeof(key_t))
+        hv = (keys[0] + (hiter * keys[1])) % bloom.length # length is in BITS
+        bloom.bitfield[hv // KEY_BITS] |= one << (hv % KEY_BITS)
```
Nice catch! So the old implementation only used 8 of the 64-bit in each bucket? What I don't fully understand is how this works with the legacy deserialization code above? In the old implementation hashes were dispersed over 8 bits and in the correct implementation over 64 bits, so don't we get completely different bit values? E.g. what may be a collision in 8 bits, may not be a collision in 64 bits.

polm replied:
In the old code, only 8 bits of each 64 bit buckets were used, but eight times as many buckets as necessary were also used. That's why this mistake didn't affect the correctness of the bloom filter. So in the correct form, the same number of bytes are actually written to, but gaps are eliminated.

If something should have used 10 bytes, for example, it would end up using 80 bytes in memory, but only 1/8 of those bytes would be written to. If those bytes are extracted and re-assembled in contiguous order, they represent the same data as in the legacy format, just without the empty gaps. 

Kind of like this:

```
A0000000 B0000000 C0000000 D0000000
... becomes ...
ABCD
```

danieldk reviewed (commented):
Nice improvements! First round of comments.

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    offset = unit_size * 3
+    hcount, length, seed = struct.unpack("<QQQ", data[0:offset])
+
+    decode_len = length // unit_size # number of units to unpack
```
Isn't `decode_len`

```suggestion
    decode_len = length // KEY_BITS
```

in both cases?

polm replied:
You are correct that it is  the same in both cases, but `length` is in terms of bytes, so `KEY_BITS` cannot be the denominator. (You can also see that using it will make the tests fail.)

polm replied:
Let me rephrase that: the units of `length` are confusing due to the bug, but because `decode_length` is used in an `unpack` call later, it needs to be expressed in terms of the container type used for unpacking, which is drawn from the legacy data.

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    # memory in the old format. In our output, length will be the bitfield
+    # length in bits.
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
```
```
buflen = decode_len if length % KEY_BITS == 0 else decode_len + 1
```

Expresses the relation more clearly. (Though I am not sure why the decode length wouldn't be increased by one as well, since we might need the extra integer there as well?)

polm replied:
As mentioned, this is not the same as `decode_len`.

`decode_len` is a number of containers to be used in the call to `unpack`. Because the data was originally serialized as a number of containers, the calculation for `decode_len` should never have a remainder. (The Windows-specific assert checks this; I could add a more general assert too.)

In contrast, the `buflen` is a buffer we're creating now, and the size is based on the number of significant bits in the incoming data. That may not be a multiple of our container size (`KEY_BITS`), which is why we have to check the modulus here and maybe increase the size.

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    # Each item in contents provides one significant byte, so we'll copy it
+    # into the containers.
+    for i in range(len(contents)):
+        block = i // sizeof(key_t)
+        idx = i % sizeof(key_t)
+        bloom.bitfield[block] |= contents[i] << (8 * idx)
```
Could you explain what is going on here? Aren't the lengths of `bloom.bitfield` and `contents` the same (`buf_len`)? Why is the index divided by the key size for the block and the shifted in bytes by the modulus?

polm replied:
This is removing the gaps from the input data to make it contiguous by repacking the data.

If the number of bits in the input is a multiple of `KEY_BITS`, then `bloom.bitfield` will be 1/8 the size of `contents` in memory. This would be awful if all of the bits in `contents` were used, but because of the bug, each item in `contents` is only using one byte, and the others are all 0 / never written to.

Pulling from the tests, it works a bit like this... (this is similar to the explanation above)

```
AA000000 BB000000 CC000000 DD000000
... converted to ...
AABBCCDD
```

In this case, `block` is which container index to go to, and `idx` is the offset within the container.

polm commented on preshed/bloom.pyx:
```diff
@@ -27,6 +29,8 @@ cdef class BloomFilter:
     Only integers are supported as keys.
     """
     def __init__(self, key_t size=(2 ** 10), key_t hash_funcs=23, uint32_t seed=0):
+        assert size > 0, "Size must be greater than zero"
```
This is not a noop, it's specifically checking that the value `!= 0` (not `> -1`). This deals with the issue described [here](https://github.com/explosion/preshed/pull/38#issuecomment-1311275420).

danieldk replied:
Sorry, misread that!

polm commented on preshed/bloom.pyx:
```diff
@@ -27,6 +29,8 @@ cdef class BloomFilter:
     """
     def __init__(self, key_t size=(2 ** 10), key_t hash_funcs=23, uint32_t seed=0):
+        assert size > 0, "Size must be greater than zero"
+        assert hash_funcs > 0, "Hash function count must be greater than zero"
```
Also not a noop.

polm commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+
+    bloom.hcount = hcount
+    bloom.length = length
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
```
You're right, I'm not sure why I thought this was necessary.

polm commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
-        return py.tostring()
+    cdef key_t pad = 0 # to differentiate new from old data format
+    cdef key_t version = 1 # hardcoded, can be incremented
+    prefix = struct.pack("<QQQQQ", pad, version, bloom.hcount, bloom.length, bloom.seed)
```
I was thinking that it would be easier to follow if the size was uniform, but I can change that.

polm commented on preshed/bloom.pyx:
```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+    for i in range(buflen):
+        bloom.bitfield[i] = contents[i]
```
I still need to look into why in more detail, but changing this to use `memcpy` causes the tests to fail.

polm replied:
Actually this does work fine, I was casting it wrong.

polm commented:
Tests were failing because the Linux image no longer has Python 3.6 available, so I have removed Python 3.6 from testing.

polm commented:
Thanks for the feedback, it's good to have this reviewed! I think I have addressed all the points that have come up at this point. The details of the repacking are definitely complicated, so let me know if that's still not clear, whether from my explanation or in code.

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +70,124 @@ cdef class BloomFilter:
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+
+    # Each item in contents provides one significant byte, so we'll copy it
```
A couple of comments use _significant bits/byte_, this confused me, because the bits that were populated in the old storage are the least significant bits/byte? E.g.:

```
        hv = (keys[0] + (hiter * keys[1])) % bloom.length
        bloom.bitfield[hv // sizeof(key_t)] |= one << (hv % sizeof(key_t))
```

E.g. if `hv=18`:

```
one << (hv % sizeof(key_t))
1 << (18 % 8)
1 << 2
00000000 00000000 00000000 00000100
```

polm replied:
Ah, that's correct - I should have used different phrasing, because I mean something like "non-zero / actually used bits" here. I'll revise that...

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +70,124 @@ cdef class BloomFilter:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    pad, ver, hcount, length, seed = struct.unpack(FORMAT, data[0:STRUCT_SIZE])
+    if pad !=0:
```
```suggestion
    if pad != 0:
```

danieldk commented on preshed/bloom.pyx:
```diff
@@ -54,40 +70,124 @@ cdef class BloomFilter:
+    if pad !=0:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    assert ver == VERSION, "Unknown serialization version"
```
This wouldn't really occur as a programming error, but if we have a future version 2 and someone loads it with an older version. Since this is a user-facing error, it should probably raise a 'normal' exception (probably `ValueError`).

danieldk commented on preshed/bloom.pyx:
```diff
@@ -109,8 +209,8 @@ cdef void bloom_add(BloomStruct* bloom, key_t item) nogil:
     for hiter in range(bloom.hcount):
-        hv = (keys[0] + (hiter * keys[1])) % bloom.length
-        bloom.bitfield[hv // sizeof(key_t)] |= one << (hv % sizeof(key_t))
+        hv = (keys[0] + (hiter * keys[1])) % bloom.length # length is in BITS
```
I wonder if we should rename `BloomStruct.length` to something like `BloomStruct.bits_length`.

I think that normally it doesn't make too much sense to put units in variable names (better to have them in types, but Cython), but I think it's really easy to confuse the units here.

Anyway, this could be a separate PR if we want this.

polm replied:
I hope that won't be ambiguous going forward, but changing the name like that is definitely a good way to avoid further issues. I won't add it to this PR for now though.

danieldk reviewed (commented):
Two small nitpicks.

polm commented:
Thanks for the feedback! I think my most recent commits should 1. clarify what I meant by "signficant" 2. raise an exception instead of using an assert for version mismatch.

svlandeg reviewed (commented):
As Daniël has reviewed this in detail, it'd be good for me to merge if he's happy.

As I understand it, care has been taken to make sure old data can still be read. Nevertheless, I wonder whether we shouldn't have this PR in a minor release of `preshed` instead of a bugfix release?

polm commented:
I think it would definitely make sense to make this a minor release of preshed.
